designer/server/src/main/scala/pl/touk/nussknacker/ui/component/ComponentsUsageHelper.scala

```diff
@@ -25,11 +27,43 @@ object ComponentsUsageHelper {
       }
     }
 
-    processesDetails
+    @tailrec
+    def mergeFragments(fragmentsProcessesDetails: List[BaseProcessDetails[ScenarioComponentsUsages]],
+                       acc: Map[ComponentId, List[(BaseProcessDetails[Unit], List[NodeId])]]): Map[ComponentId, List[(BaseProcessDetails[Unit], List[NodeId])]] =
+      fragmentsProcessesDetails match {
+        case Nil => acc
+        case fragmentProcessesDetails :: t => {
+          val componentIdUsages = toComponentIdUsages(fragmentProcessesDetails).toMap
```

Comment:
Can't you just compute it once before this method? It will be faster and will have only arguments that are changeable. Also let's rename it to sth with "fragment" in name

Reply:
Note that I compute it per `fragment`, not `fragments`.